### ISSUE TITLE
When I make a request which includes a TE header field indicating "trailers，The response header only returns the key of metadata, but does not return the value of Trailer metadata

### ISSUE THREAD

**Issue by tocatch** [grpc-ecosystem/grpc-gateway#2067]
## 🐛 Bug Report

When I make a request which includes a TE header field indicating "trailers，The response header only returns the key of metadata, but does not return the value of Trailer metadata

Means in the file of error. go , the method of handleForwardResponseTrailerHeader (w, md) to take effect, but handleForwardResponseTrailer (w, md) do not take effect

When I will be in handleForwardResponseTrailer (w, md) this line of code in w.WriteHeader (st) before,It takes effect

<img width="680" alt="image" src="https://user-images.githubusercontent.com/24827471/112952553-32866d00-916f-11eb-91ac-d8b9dace72cf.png">

## To Reproduce

(Write your steps here:)

1. Start the project
2. Set a header which includes a TE header field indicating "trailers
3. To observe the response headers

## Expected behavior

The response header returns the key of metadata, and return the value of Trailer metadata

## Actual Behavior

The response header only returns the key of metadata, but does not return the value of Trailer metadata

## Your Environment

grpc-gateway v2

go 1.14.4

mac os

**Comment by tocatch** [comment]
> Let me rephrase so that I make sure I understand you correctly. You're saying that the trailers values are not sent correctly and calling `handleForwardResponseTrailer(w, md)` before `w.WriteHeader(st)` will fix it. Looking at the current code, it seems correct and sending trailers is done as it's expected. You can take a look [at this example](https://golang.org/pkg/net/http/#example_ResponseWriter_trailers).
> 
> Please remember that trailer headers are sent after the body. In Go, you can read them from [`Response.Trailer` field](https://golang.org/pkg/net/http/#Response).
> 
> I will try to replicate with a simple example and verify if this works as expected in the current master.

**Comment by tocatch** [comment]
> Let me rephrase so that I make sure I understand you correctly. You're saying that the trailers values are not sent correctly and calling `handleForwardResponseTrailer(w, md)` before `w.WriteHeader(st)` will fix it. Looking at the current code, it seems correct and sending trailers is done as it's expected. You can take a look [at this example](https://golang.org/pkg/net/http/#example_ResponseWriter_trailers).
> 
> Please remember that trailer headers are sent after the body. In Go, you can read them from [`Response.Trailer` field](https://golang.org/pkg/net/http/#Response).
> 
> I will try to replicate with a simple example and verify if this works as expected in the current master.

Your understanding is right, and I hope to get the result of your example. Thank you very much

**Comment by adambabik** [comment]
Hi @tocatch. I wrote a unit test to in order to reproduce the bug but it turned out that it works as intendet.

```go
func TestDefaultHTTPErrorHandler_Trailers(t *testing.T) {
	const (
		trailerHeaderName       = "Header-A"
		trailerHeaderValue      = "a-header-value"
		resultTrailedHeaderName = runtime.MetadataTrailerPrefix + trailerHeaderName
	)

	serverMetadata := runtime.ServerMetadata{
		TrailerMD: metadata.New(map[string]string{trailerHeaderName: trailerHeaderValue}),
	}
	ctx := runtime.NewServerMetadataContext(context.Background(), serverMetadata)
	mux := runtime.NewServeMux()
	marshaler := &runtime.JSONPb{}
	w := httptest.NewRecorder()
	req, _ := http.NewRequest("", "", nil) // Pass in an empty request to match the signature
	req.Header.Add("TE", "trailers")

	runtime.HTTPError(ctx, mux, marshaler, w, req, fmt.Errorf("example error"))

	result := w.Result()
	if got, want := result.Header.Get("Trailer"), resultTrailedHeaderName; got != want {
		t.Errorf("headers do not include expected Trailer header")
	}
	if got, want := result.Trailer.Get(resultTrailedHeaderName), trailerHeaderValue; got != want {
		t.Errorf("trailer header not found")
	}
}
```

I am not sure what's your explanation but you wrote "3. To observe the response headers". As you see in the test, the trailer header is only communicated in the actual response headers and it arrives later. If you use curl, you can see that using `-v --raw` flags.

**Comment by tocatch** [comment]
Thank you very much for your time to help me verify this problem. This problem is a problem of my local environment. I mistakenly thought that metadata was put into the header of HTTP response, but actually it was put into the trailers of response. My problem has been solved. Thanks again
